rfc/store-deliver.md

```diff
@@ -0,0 +1,278 @@
+  "nb": {
+    "link": "bag...",
+    "size": 1234
+  }
+}
+```
+
+If the service is receptive to store this data, a presigned URL is returned in the receipt of `store/add` invocation, so that the client can store this content. The presigned URL includes a hash derived from the content CID to guarantee the written bytes match that CID. A receipt example:
+
+```json
+{
+  "ran": "bafy...storeAdd",
+  "out": {
+    "ok": {
+      "status": "upload",
+      "url": "https://...",
+      "link": "bag...",
+      "allocated": 1234
+    }
+  },
+  "fx": {
+    "join": null,
+    "fork": []
+  },
+}
+```
+
+Note that content requested to store, may already be stored. If that is the case, the service may issue receipt with `"status": "done"`.
+
+On their schedule, the agent can post the bytes using the presigned URL. However, both the agent and the service never sign a receipt stating that this content is now stored. This is where `store/deliver` can help.
+
+While not relevant for verifiability, it is also important pointing out that the service MAY need to run some data computation, issue billing events or write indexes for retrievability. Today, web3.storage operated store service relies on "S3 Bucket events" to achieve this, but it should not be assumed that all stores web3.storage will rely on will provide such features.
+
+## store/delivery capability
+
+Aiming to introduce verifiability into content storing, as well as to pave the way for a future proof of delivery, `store/deliver` capability is proposed. It would also make services not tied with store put events to execute side effects.
+
+To offer verifiability, `store/add` invocation should be associated with `store/deliver` via [effect]s. `store/deliver` would be invoked by the client to notify the service that content was stored, and the the service should also sign a confirmation that it is true.
+
+### store/delivery of not previously stored content
+
+Taking into account the above, we can look at the following flow walkthrough and diagram:
+
+1. client invokes `store/add`
+
+```json
+{
+  "can": "store/add",
+  "with": "did:key:abc...",
+  "nb": {
+    "link": "bag...",
+    "size": 1234
+  }
+}
+```
+
+2. service issues `store/add` receipt with presigned URL and effect for `store/deliver` that may be invoked in the future by the service
+
+```json
+{
+  "ran": "bafy...storeAdd",
+  "out": {
+    "ok": {
+      "status": "upload",
+      "url": "https://...",
+      "link": "bag...",
+      "allocated": 1234
+    }
+  },
+  "fx": {
+    "join": { "/": "bafy...storeDeliverService" },
+    "fork": []
+  },
+}
+```
+
+3. client posts the content to the provided presigned URL.
+
+4. client invokes `store/deliver`
+
+```json
+{
+  "can": "store/deliver",
+  "with": "did:key:abc...",
+  "nb": {
+    "link": "bag...",
+    "url": "https://..."
+  }
+}
+```
+
+5. service verifies if content was delivered by user, and queues a self invocation of `store/deliver` that is also provided as an effect.
```

Comment:
I'd like this to be less complicated - in new invocation spec effects are CIDs to _tasks_ not _invocations_ so the server could issue this next task in the receipt and the client can invoke to complete the task and then we're done.

Reply:
I agree with this, that said there is still needs to be a pending task awaiting for `store/deliver` receipt that will issue receipt from us that workflow was complete. I tried to show this in the flowchart in https://github.com/web3-storage/RFC/pull/10/files#r1514851900